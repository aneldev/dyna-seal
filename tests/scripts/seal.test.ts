import {
  createDynaSeal,
  validateDynaSeal,
  ESealValidationResolution,
} from "../../src";

describe('seal', () => {
  const product = {
    productId: '9824805246245',
    passenger: {
      firstName: 'John',
      lastName: 'Smith',
    },
    price: {
      value: 1200.34,
      currency: 'eur',
    },
    seal: '',
  };
  type IProduct = typeof product;

  const createFingerprint = (product: IProduct): string =>
    [
      product.productId,
      product.passenger.firstName,
      product.passenger.lastName,
    ].join('/');

  test('Valid seal', () => {
    const expiresAt = Date.now() + 20000;
    const sealedProduct = createDynaSeal<IProduct>({
      data: product,
      createFingerprint,
      encryptionKey: 'sssh-secret',
      expiresAt,
    });

    const validation = validateDynaSeal<IProduct>({
      data: sealedProduct,
      createFingerprint,
      encryptionKey: 'sssh-secret',
    });

    expect(validation.valid).toBeTruthy();
    expect(validation.expiresAt).toBe(expiresAt);
    expect(validation.expiresAt).toBeGreaterThan(Date.now());
  });

  test('Invalid, with NO_SEAL error', () => {
    const validation = validateDynaSeal<IProduct>({
      data: product,
      createFingerprint,
      encryptionKey: 'sssh-secret',
    });

    // Will be invalid, since the product object has not seal
    expect(validation.valid).toBeFalsy();
    expect(validation.resolution).toBe(ESealValidationResolution.NO_SEAL);
    expect(validation.expiresAt).toBe(-1);
  });

  test('Invalid, with DECRYPTION_ERROR error', () => {
    const expiresAt = Date.now() + 20000;
    const sealedProduct = createDynaSeal<IProduct>({
      data: product,
      createFingerprint,
      encryptionKey: 'sssh-secret',
      expiresAt,
    });

    const validation = validateDynaSeal<IProduct>({
      data: sealedProduct,
      createFingerprint,
      encryptionKey: 'sssh-secret--WRONG-KEY',
    });

    expect(validation.valid).toBeFalsy();
    expect(validation.resolution).toBe(ESealValidationResolution.DECRYPTION_ERROR);
    expect(validation.expiresAt).toBe(-1);
  });

  test('Invalid, with DATA_MUTATED error', () => {
    const expiresAt = Date.now() + 20000;
    const sealedProduct = createDynaSeal<IProduct>({
      data: product,
      createFingerprint,
      encryptionKey: 'sssh-secret',
      expiresAt,
    });

    const validation = validateDynaSeal<IProduct>({
      data: {
        ...sealedProduct,
        productId: 'Diff--product-id--43524',
      },
      createFingerprint,
      encryptionKey: 'sssh-secret',
    });

    expect(validation.valid).toBeFalsy();
    expect(validation.expiresAt).toBe(expiresAt);
    expect(validation.resolution).toBe(ESealValidationResolution.DATA_MUTATED);
  });

  test('Invalid, with EXPIRED error', () => {
    const expiresAt = Date.now() + -20000;
    const sealedProduct = createDynaSeal<IProduct>({
      data: product,
      createFingerprint,
      encryptionKey: 'sssh-secret',
      expiresAt,
    });

    const validation = validateDynaSeal<IProduct>({
      data: {
        ...sealedProduct,
      },
      createFingerprint,
      encryptionKey: 'sssh-secret',
    });

    expect(validation.valid).toBeFalsy();
    expect(validation.expiresAt).toBe(expiresAt);
    expect(validation.resolution).toBe(ESealValidationResolution.EXPIRED);
  });
});
