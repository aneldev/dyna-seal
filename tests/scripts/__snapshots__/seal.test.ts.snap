// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seal Valid seal 1`] = `
Object {
  "passenger": Object {
    "firstName": "John",
    "lastName": "Smith",
  },
  "price": Object {
    "currency": "eur",
    "value": 1200.34,
  },
  "productId": "9824805246245",
  "seal": "U2FsdGVkX181Ih63aVj+jD1N/nDKmHxiLD9adM1TyMYKEjrNIWRpnEtS4h2OwDxi9shFWIbBOjT5ikvXps4u1yAB6RfXaLv5SbBt33E4w9fj9qfbfO2gNb1GJjBSDcaZ",
}
`;
